File: packages/kernel/src/kernel-info.ts
import fs from "fs/promises";
import path from "path";
import os from "os";

export interface KernelInfo {
  port: number;
  token: string;
  pid: number;
  startedAt: string;
  scratchDir: string;
}

const CRONA_DIR = path.join(os.homedir(), ".crona");
const KERNEL_INFO_FILE = path.join(CRONA_DIR, "kernel.json");
export const CRONA_SCRATCH_DIR = path.join(CRONA_DIR, "scratch");

/**
 * Ensure ~/.crona exists with safe permissions
 */
async function ensureDir() {
  await fs.mkdir(CRONA_DIR, { recursive: true, mode: 0o700 });
  await fs.mkdir(CRONA_SCRATCH_DIR, { recursive: true, mode: 0o700 });
}

/**
 * Write kernel connection info
 * Overwrites existing file atomically
 */
export async function writeKernelInfo(info: Omit<KernelInfo, "scratchDir">): Promise<void> {
  await ensureDir();

  const tempFile = `${KERNEL_INFO_FILE}.tmp`;

  await fs.writeFile(
    tempFile,
    JSON.stringify({
      ...info,
      scratchDir: CRONA_SCRATCH_DIR,
    }, null, 2),
    { mode: 0o600 }
  );

  await fs.rename(tempFile, KERNEL_INFO_FILE);
}

/**
 * Read kernel info if present
 */
export async function readKernelInfo(): Promise<KernelInfo | null> {
  try {
    const raw = await fs.readFile(KERNEL_INFO_FILE, "utf8");
    return JSON.parse(raw) as KernelInfo;
  } catch {
    return null;
  }
}

/**
 * Remove kernel info (on shutdown / crash recovery)
 */
export async function clearKernelInfo(): Promise<void> {
  try {
    await fs.unlink(KERNEL_INFO_FILE);
  } catch (err) {
    console.error("Failed to clear kernel info:", err);
  }
}

File: kernel/internal/runtime/kernel_info.go
package runtime

import (
	"encoding/json"
	"os"

	sharedtypes "crona/shared/types"
)

func WriteKernelInfo(paths Paths, info sharedtypes.KernelInfo) error {
	tmpPath := paths.InfoFile + ".tmp"

	body, err := json.MarshalIndent(info, "", "  ")
	if err != nil {
		return err
	}

	if err := os.WriteFile(tmpPath, body, FilePerm()); err != nil {
		return err
	}

	return os.Rename(tmpPath, paths.InfoFile)
}

func ClearKernelInfo(paths Paths) error {
	err := os.Remove(paths.InfoFile)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

func ReadKernelInfo(paths Paths) (*sharedtypes.KernelInfo, error) {
	body, err := os.ReadFile(paths.InfoFile)
	if err != nil {
		return nil, err
	}

	var info sharedtypes.KernelInfo
	if err := json.Unmarshal(body, &info); err != nil {
		return nil, err
	}

	return &info, nil
}
